src/Lykke.Service.HftInternalService.Services/ApiKeyService.cs

```diff
@@ -1,40 +1,39 @@
 ﻿using System;
 using System.Linq;
 using System.Threading.Tasks;
-using Lykke.Service.HftInternalService.Core;
+using JetBrains.Annotations;
+using Lykke.Service.HftInternalService.Contracts.Events;
 using Lykke.Service.HftInternalService.Core.Domain;
 using Lykke.Service.HftInternalService.Core.Services;
-using Microsoft.Extensions.Caching.Distributed;
+using ApiKey = Lykke.Service.HftInternalService.Core.Domain.ApiKey;
 
 namespace Lykke.Service.HftInternalService.Services
 {
     public class ApiKeyService : IApiKeyService
     {
-        private readonly IDistributedCache _distributedCache;
-        private readonly CacheSettings _cacheSettings;
+        private readonly IApiKeyPublisher _apiKeyPublisher;
         private readonly IRepository<ApiKey> _apiKeyRepository;
 
-        public ApiKeyService(IDistributedCache distributedCache, CacheSettings cacheSettings, IRepository<ApiKey> orderStateRepository)
+        public ApiKeyService([NotNull] IRepository<ApiKey> orderStateRepository, [NotNull] IApiKeyPublisher apiKeyPublisher)
         {
-            _cacheSettings = cacheSettings ?? throw new ArgumentNullException(nameof(cacheSettings));
-            _distributedCache = distributedCache ?? throw new ArgumentNullException(nameof(distributedCache));
+            _apiKeyPublisher = apiKeyPublisher ?? throw new ArgumentNullException(nameof(apiKeyPublisher));
             _apiKeyRepository = orderStateRepository ?? throw new ArgumentNullException(nameof(orderStateRepository));
         }
 
         public async Task<ApiKey> GenerateApiKeyAsync(string clientId, string walletId)
         {
             var apiKey = Guid.NewGuid();
-            var apiKeyAsString = apiKey.ToString();
-            await _distributedCache.SetStringAsync(GetCacheKey(apiKeyAsString), walletId);
+
             var existedApiKey = await _apiKeyRepository.Get(x => x.WalletId == walletId && x.ValidTill == null);
             if (existedApiKey != null)
             {
-                await _distributedCache.RemoveAsync(GetCacheKey(existedApiKey.Id.ToString()));
                 existedApiKey.ValidTill = DateTime.UtcNow;
                 await _apiKeyRepository.Update(existedApiKey);
+                await PublishNotification(existedApiKey);
             }
             var key = new ApiKey { Id = apiKey, ClientId = clientId, WalletId = walletId };
             await _apiKeyRepository.Add(key);
+            await PublishNotification(key);
```

Comment:
If PublishNotification(key); throws an exception the key stays in _apiKeyRepository is it normal state?

Reply:
It is ok to save key no matter of publishing result. But I'll add a try/catch block for PublishNotification() method.